### PR TITLE
TE-3.7: update with changes from #1135

### PR DESCRIPTION
**Pull request by LimeHat**
Implements changes described in #1135 by @xw-g 

- Implement TODOs
- Add implicit in-place add scenario.
- Add implicit in-place remove scenario.
- Update implicit in-place replace scenario.

---
<sup>
This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia’s intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind.</sup>

### CODE REVIEW

File: feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
Comment (xw-g) [inline feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go:198]: Can you help update the comment to be more aligned? `blablabla to dstPfx via a VIP`.

File: feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
Comment (xw-g) [inline feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go:54]: A better name would be `virtualIP`?.

File: feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
Comment (xw-g) [inline feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go:52]: If you agree, then this would be "virtualIPNHGID".

**Review by xw-g — COMMENTED** [review]
* Please also help update the README of the OTG test.

**Comment by xw-g** [comment]
@mojiiba The code changes are to implement #1135, and the code looks good to me. Can you help take another look if any concerns blocking the merge?

For additional "deviations" needed by your team, please help open another PR later. Thanks.

**Comment by mojiiba** [comment]
> @mojiiba The code changes are to implement #1135 (to close some testing gaps), and the code looks good to me. Can you help take another look if any concerns blocking the merge?
> 
> For additional "deviations" needed by your team, please help open another PR later. Thanks.

sure, @manan-patel is reviewing it.

File: feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
Comment (manan-patel) [inline feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go:156]: Why are we not using the existing function or enhancing it ?
featureprofiles/internal/gribi/gribi.go

File: feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
Comment (manan-patel) [inline feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go:176]: Why are we not using the existing function or enhancing it ?
featureprofiles/internal/gribi/gribi.go

File: feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
Comment (manan-patel) [inline feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go:236]: Can we enhance AddIPv4 under featureprofiles/internal/gribi/gribi.go?

File: feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
Comment (LimeHat) [reply feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go:156]: The existing code helper doesn't cover this use case as you know + this is the way this test was written from the beginning.

We can update the helper and rewrite the test indeed, I would prefer to do this as a separate PR though. Would that be ok with you?

**Comment by manan-patel** [comment]
@LimeHat, we will need an existing deviation (GRIBIMACOverrideWithStaticARP), Can you please integrate the below code


git diff
diff --git a/feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go b/feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
index be0981c..543c06f 100644
--- a/feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
+++ b/feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
@@ -86,6 +86,17 @@ var (
                IPv4Len: 30,
        }
 
+       dutPort2DummyIP = attrs.Attributes{
+               Desc:    "DUT Port 2",
+               IPv4:    "192.0.2.21",
+               IPv4Len: 30,
+       }
+       dutPort3DummyIP = attrs.Attributes{
+               Desc:    "DUT Port 3",
+               IPv4:    "192.0.2.41",
+               IPv4Len: 30,
+       }
+
        atePort1 = attrs.Attributes{
                Name:    "atePort1",
                IPv4:    "192.0.2.2",
@@ -135,12 +146,31 @@ func TestBaseHierarchicalNHGUpdate(t *testing.T) {
        dutP2 := dut.Port(t, "port2").Name()
        dutP3 := dut.Port(t, "port3").Name()
        t.Logf("Adding gribi routes and validating traffic forwarding via port %v and NH ID %v", dutP2, p2NHID)
-       addInterfaceRoute(ctx, t, gribic, p2NHID, dutP2)
+
+       if deviations.GRIBIMACOverrideWithStaticARP(ondatra.DUT(t, "dut")) {
+               d := gnmi.OC()
+               p2 := dut.Port(t, "port2")
+               p3 := dut.Port(t, "port3")
+               gnmi.Update(t, dut, d.Interface(p2.Name()).Config(), dutPort2DummyIP.NewOCInterface(p2.Name()))
+               gnmi.Update(t, dut, d.Interface(p3.Name()).Config(), dutPort3DummyIP.NewOCInterface(p3.Name()))
+               gnmi.Update(t, dut, d.Interface(p2.Name()).Config(), configStaticArp(t, p2, "192.0.2.22", pMAC))
+               gnmi.Update(t, dut, d.Interface(p3.Name()).Config(), configStaticArp(t, p3, "192.0.2.42", pMAC))
+       }
+
+       if deviations.GRIBIMACOverrideWithStaticARP(ondatra.DUT(t, "dut")) {
+               addInterfaceRoute(ctx, t, gribic, p2NHID, dutP2, "192.0.2.22")
+       } else {
+               addInterfaceRoute(ctx, t, gribic, p2NHID, dutP2)
+       }
        addDestinationRoute(ctx, t, gribic)
        validateTrafficFlows(t, ate, []*ondatra.Flow{p2flow}, []*ondatra.Flow{p3flow}, nil, pMACFilter)
 
        t.Logf("Adding a new NH via port %v with ID %v", dutP3, p3NHID)
-       addNH(ctx, t, gribic, p3NHID, dutP3, pMAC)
+       if deviations.GRIBIMACOverrideWithStaticARP(ondatra.DUT(t, "dut")) {
+               addNH(ctx, t, gribic, p3NHID, dutP3, pMAC, "192.0.2.42")
+       } else {
+               addNH(ctx, t, gribic, p3NHID, dutP3, pMAC)
+       }
        t.Logf("Performing implicit in-place replace with two next-hops (NH IDs: %v and %v)", p2NHID, p3NHID)
        addNHG(ctx, t, gribic, interfaceNHGID, []uint64{p2NHID, p3NHID})
        validateTrafficFlows(t, ate, nil, nil, []*ondatra.Flow{p2flow, p3flow}, pMACFilter)
@@ -153,9 +183,12 @@ func TestBaseHierarchicalNHGUpdate(t *testing.T) {
 }
 
 // addNH adds a GRIBI NH with a FIB ACK confirmation via Modify RPC
-func addNH(ctx context.Context, t *testing.T, gribic *fluent.GRIBIClient, id uint64, intf, mac string) {
+func addNH(ctx context.Context, t *testing.T, gribic *fluent.GRIBIClient, id uint64, intf, mac string, nhip ...string) {
        nh := fluent.NextHopEntry().WithNetworkInstance(*deviations.DefaultNetworkInstance).
                WithIndex(id).WithInterfaceRef(intf).WithMacAddress(mac)
+       if len(nhip) > 0 {
+               nh = nh.WithIPAddress(nhip[0])
+       }
        gribic.Modify().AddEntry(t, nh)
        if err := awaitTimeout(ctx, gribic, t, time.Minute); err != nil {
                t.Fatalf("Await got error for entries: %v", err)
@@ -233,9 +266,12 @@ func addDestinationRoute(ctx context.Context, t *testing.T, gribic *fluent.GRIBI
 
 // addInterfaceRoute creates a GRIBI route that points to the egress interface defined by id,
 // port, and nhip.
-func addInterfaceRoute(ctx context.Context, t *testing.T, gribic *fluent.GRIBIClient, id uint64, port string) {
+func addInterfaceRoute(ctx context.Context, t *testing.T, gribic *fluent.GRIBIClient, id uint64, port string, nhip ...string) {
        inh := fluent.NextHopEntry().WithNetworkInstance(*deviations.DefaultNetworkInstance).
                WithIndex(id).WithInterfaceRef(port).WithMacAddress(pMAC)
+       if len(nhip) > 0 {
+               inh = inh.WithIPAddress(nhip[0])
+       }
        inhg := fluent.NextHopGroupEntry().WithNetworkInstance(*deviations.DefaultNetworkInstance).
                WithID(interfaceNHGID).AddNextHop(id, 1)
        ipfx := fluent.IPv4Entry().WithNetworkInstance(*deviations.DefaultNetworkInstance).
@@ -435,3 +471,13 @@ func validateTrafficFlows(t *testing.T, ate *ondatra.ATEDevice, good, bad, lb []
                }
        }
 }
+
+func configStaticArp(t *testing.T, p *ondatra.Port, ipv4addr string, macAddr string) *oc.Interface {
+       i := &oc.Interface{Name: ygot.String(p.Name())}
+       i.Type = oc.IETFInterfaces_InterfaceType_ethernetCsmacd
+       s := i.GetOrCreateSubinterface(0)
+       s4 := s.GetOrCreateIpv4()
+       n4 := s4.GetOrCreateNeighbor(ipv4addr)
+       n4.LinkLayerAddress = ygot.String(macAddr)
+       return i
+}

**Comment by manan-patel** [comment]
@LimeHat We need the following deviation "GRIBIMACOverrideWithStaticARP", can you please add it.
[TE-3.7-deviation.log](https://github.com/openconfig/featureprofiles/files/11174513/TE-3.7-deviation.log)

File: feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go
Comment (manan-patel) [reply feature/gribi/ate_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go:156]: Yes, I saw the TC history, it was written this way from the beginning. Please go ahead, we can address it in a separate PR.

**Comment by LimeHat** [comment]
Thanks for the review @xw-g & @manan-patel , updated according to the comments/discussion.

**Comment by xw-g** [comment]
@manan-patel any further concern? Please help approve this PR if there is no further concerns. 

Note that when one vender needs a deviation I strongly prefer another incremental PR to be filed by that vendor to iterate the code, instead of asking the PR author to do it. Since @LimeHat already did it, I'm ok to proceed. 

Also. @cprabha any concerns with this PR? Thanks.

**Review by xw-g — APPROVED** [review]

**Comment by manan-patel** [comment]
@xw-g 

Not all the provided diffs are merged by @LimeHat, this TC will start failing once PR is merged for some vendors. To address that we should submit another deviation PR right away

Adding @vosipchu @prinikasn @guoshiuan to keep in loop

**Comment by LimeHat** [comment]
@manan-patel could you elaborate on what is missing?
I think the only part I didn't add is the "dummy IP" configuration, but I'm not sure what is the purpose of using it (if you only need a static ARP entry?)
Or it has to be a static arp pointing to a new IP address?

**Comment by manan-patel** [comment]
@LimeHat 

As a part of the deviation, if we just configure static arp on port 2 and port 3 it will end up not establishing a connection between DUT<>ATE.
By having a dummy ip, we configure static arp to those addresses and use them for gribi programming, while the link between DUT<>ATE gets resolved by dynamic arp.

**Comment by LimeHat** [comment]
Alright, I added the dummy ip cfg as well.

**Comment by xw-g** [comment]
@manan-patel I prefer your team to file a quick PR to iterate, once this is merged. I'm not expecting one vendor to know all details about another vendor's deviation behavior and addressed it in the same PR. 

So, do you have other non-deviation related concerns? ok to merge?

**Comment by manan-patel** [comment]
@LimeHat Thank you for addressing it
@xw-g Acknowledged

**Review by manan-patel — APPROVED** [review]

**Review by xw-g — APPROVED** [review]

**Review by sachendras — APPROVED** [review]

**Pull request merged**
